src/serializers/ob_type.rs

```diff
@@ -245,11 +273,68 @@ impl ObTypeLookup {
             let meta_type = unsafe { (*type_ptr).ob_type };
             #[cfg(any(not(PyPy), Py_3_9))]
             let meta_type = unsafe { (*type_ptr).ob_base.ob_base.ob_type };
-            meta_type as usize == self.enum_type
+            meta_type as usize == self.enum_
         } else {
             false
         }
     }
+
+    /// If our logic for finding types by recursively checking `tp_base` fails, we fallback to this which
+    /// uses `isinstance` thus supporting both mixins and classes that implement `__instancecheck__`.
+    /// We care about order here since:
+    /// 1. we pay a price for each `isinstance` call
+    /// 2. some types are subclasses of others, e.g. `bool` is a subclass of `int`
+    /// hence we put common types first
+    fn fallback_isinstance(&self, value: &PyAny) -> ObType {
+        let py = value.py();
+        if PyBool::is_type_of(value) {
+            ObType::Bool
+        } else if PyInt::is_type_of(value) {
+            ObType::IntSubclass
+        } else if PyFloat::is_type_of(value) {
+            ObType::FloatSubclass
+        } else if PyString::is_type_of(value) {
+            ObType::StrSubclass
+        } else if PyBytes::is_type_of(value) {
+            ObType::Bytes
+        } else if PyByteArray::is_type_of(value) {
+            ObType::Bytearray
+        } else if PyList::is_type_of(value) {
+            ObType::List
+        } else if PyTuple::is_type_of(value) {
+            ObType::Tuple
+        } else if PyDict::is_type_of(value) {
+            ObType::Dict
+        } else if PySet::is_type_of(value) {
```

Comment:
There are some types which have inheritance set as a bitflag on the type opbejct:

https://github.com/python/cpython/blob/52fbcf61b5a70993c2d32332ff0ad9f369d968d3/Include/object.h#L552-L559

```
#define Py_TPFLAGS_LONG_SUBCLASS        (1UL << 24)
#define Py_TPFLAGS_LIST_SUBCLASS        (1UL << 25)
#define Py_TPFLAGS_TUPLE_SUBCLASS       (1UL << 26)
#define Py_TPFLAGS_BYTES_SUBCLASS       (1UL << 27)
#define Py_TPFLAGS_UNICODE_SUBCLASS     (1UL << 28)
#define Py_TPFLAGS_DICT_SUBCLASS        (1UL << 29)
#define Py_TPFLAGS_BASE_EXC_SUBCLASS    (1UL << 30)
#define Py_TPFLAGS_TYPE_SUBCLASS        (1UL << 31)
```

`isinstance` for these types should hopefully be fast, you may want to put them first.